scRNA-seq/07-gene_set_enrichment_analysis.Rmd

```diff
@@ -0,0 +1,352 @@
+# data frame of results
+filtered_markers_df %>%
+  dplyr::filter(gene_symbol %in% duplicated_gene_symbols) %>%
+  dplyr::arrange(gene_symbol)
+```
+
+Now we're ready to prep our pre-ranked list for GSEA.
+
+### Pre-ranked list
+
+The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+This is step 1 -- gene-level statistics.
+
+Here, we're using the summary log fold change, which summarizes of all of the log fold changes from the pairwise comparisons between cluster 1 and other clusters into a single value ([ref](https://rdrr.io/bioc/scran/man/combineMarkers.html)). 
+As such, this statistic gives us information about the magnitude and directionality of a gene's expression in cluster 1 vs. all other clusters.
+
+```{r lfc_vector}
+lfc_vector <- filtered_markers_df %>%
+  # Extract a vector of `summary.logFC` named by `gene_symbol`
+  dplyr::pull(summary.logFC, name = gene_symbol)
+lfc_vector <- sort(lfc_vector, decreasing = TRUE)
+```
+
+Let's look at the top ranked values.
+
+```{r head_lfc, live = TRUE}
+# Look at first entries of the log fold change vector
+head(lfc_vector)
+```
+
+And the bottom of the list.
+
+```{r tail_lfc, live = TRUE}
+# Look at the last entries of the log fold change vector
+tail(lfc_vector)
+```
+
+## Run GSEA
+
+Now for the analysis!
+
+We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).
+
+```{r run_gsea}
+gsea_results <- GSEA(geneList = lfc_vector,  # ordered ranked gene list
+                     minGSSize = 25,  # minimum gene set size
+                     maxGSSize = 500,  # maximum gene set set
+                     pvalueCutoff = 0.05,
+                     pAdjustMethod = "BH",  # correction for multiple hypothesis testing
+                     TERM2GENE = dplyr::select(hs_immunologic_df,
+                                               gs_name,
+                                               gene_symbol))
+```
+Let's take a look at the GSEA results.
+
+```{r view_gsea, live = TRUE, eval = FALSE}
+View(gsea_results@result %>%
+       dplyr::arrange(dplyr::desc(NES))
+)
+```
+
+Normalized enrichment scores (NES) are enrichment scores that are scaled to make gene sets that contain different number of genes comparable.
+
+Let's write these results to file.
+
+```{r write_gsea}
+gsea_results@result %>% readr::write_tsv(output_file)
+```
+
+### Visualizing GSEA results
+
+We can visualize GSEA results for individual pathways or gene sets using `enrichplot::gseaplot()`.
+Let's take a look at 3 different pathways -- one with a highly positive NES, one with a highly negative NES, and one that was not a significant result -- to get more insight into how ES are calculated.
+
+#### Highly Positive NES
+
+Let's take look at a pathway with a highly positive NES (`GSE29618_BCELL_VS_PDC_UP`) using a GSEA plot.
```

Comment:
Do we want to say anything about what these pathways "are"? Maybe a link to https://www.gsea-msigdb.org/gsea/msigdb/cards/GSE29618_BCELL_VS_PDC_UP, etc.

Reply:
Added quotes of the descriptions for the highly positive and highly negative examples from MSigDB + used the link for the non-significant result!